feat: pino logging

trentm opened the pull request:
This switches the APM agent to use pino for structured logging. Visible
changes:

- Default logging output is now JSON in ecs-logging format to stdout,
  rather than just a plain message to stdout or stderr (warn and above).
- Custom loggers are still supported and there shouldn't be any breakage
  there. This is done by wrapping the custom logger in a pino logger
  that will just pass the log message through. This means that when we
  add structure log fields (for example "foo" in `log.info({foo:
  bar}, 'my message')`), then users of a custom logger will not see
  those structured fields. I think that is fine.
- Added support for `ELASTIC_APM_LOGGER=false` to ignore a given custom logger to be used for "Debug mode" troubleshooting logging... i.e. to avoid not getting structured log fields in debug output from users for the (hopefully rare) cases where they are using a custom logger.

Note that I want this improved logging as a part of the
https://github.com/elastic/apm-nodejs-http-client/issues/136 work.

TODO:
- [x] The logLevel config var now *applies* to a given custom logger.
  Discuss this. Custom loggers are wrapped in a pino logger that
  writes the log message to the custom logger. The logLevel is on that
  pino logger wrapper. I think I should change this to set the wrapper
  to 'trace' level and *not* use `logLevel` as before... otherwise the
  user has *two* confusing log levels to coordinate.
- [x] Drop the 'pid' and 'hostname' default fields as discussed in #1430
  that this PR is obsoleting. I need to fix a bug in
  @elastic/ecs-format-pino for this.
- [x] Check/handle this case:
  `this.logger.warn(new Error('cannot flush agent before it is started'))`
- [x] Do we need `crlf: platform() === 'win32',` from the prev PR?
  I am inclined to *no* unless there is specific argument that it is required. My memory from Windows land was that basically everything except Notepad was fine with LF line endings. If that is the only reason, then no point in special casing here.
- [x] Update troubleshooting doc for logging changes. Including turning
  *off* a custom logger than can hide data.

Things to do after this PR:
- pass the logger to the http client
- change over lots of the current logging statements to use structured
  data, including a 'span' serializer (as in PR #1430)

<!--

Replace this comment with a description of what's being changed by this PR.

If this PR should close an issue, please add one of the magic keywords
(e.g. Fixes) followed by the issue number. For more info see:
https://help.github.com/articles/closing-issues-using-keywords/

-->

### Checklist

<!-- Potential tasks related to a new PR. Remove tasks that are not relevant -->

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings (N/A)
- [x] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-nodejs_2001.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #2001 opened`



* Start Time: 2021-03-16T00:35:48.501+0000


* Duration: 10 min 7 sec


* Commit: b4a75400d5ef0f11eadd90e0097daaa9f7c99840



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 16614  |
| Skipped      | 0 |
| Total        | 16614   |


#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2001/buildTimeGraph/png)


![Image of Tests](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2001/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![38](https://img.shields.io/badge/38%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


> Show only the first 10 steps failures

  
  
  
  
##### `Run Tests`
<ul>
<li>Took 3 min 45 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/395/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 0 min 28 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/678/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 3 min 39 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/400/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 0 min 30 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/657/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 2 min 22 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/355/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 0 min 24 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/517/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 2 min 26 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/401/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Run Tests`
<ul>
<li>Took 0 min 26 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/572/log/?start=0">here</a></li>
<li>Description: <code>.ci/scripts/test.sh "8.6" "" "false" </code></l1>
</ul>
  
  
  
  
##### `Windows Batch Script`
<ul>
<li>Took 0 min 7 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/566/log/?start=0">here</a></li>
<li>Description: <code>node test/test.js</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/1/steps/558/log/?start=0">here</a></li>
<li>Description: <code>hudson.AbortException: script returned exit code 1</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-03-16T00:45:43.398Z] /tmp/.npm-global/bin/tap-xunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-03-16T00:45:43.398Z] /tmp/.npm-global/bin/txunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-03-16T00:45:43.398Z] + tap-xunit@2.4.1
[2021-03-16T00:45:43.398Z] added 21 packages from 21 contributors in 2.75s
[2021-03-16T00:45:43.453Z] Recording test results
[2021-03-16T00:45:44.988Z] [Checks API] No suitable checks publisher found.
[2021-03-16T00:45:45.037Z] Running in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001/src/github.com/elastic/apm-agent-nodejs
[2021-03-16T00:45:45.085Z] [INFO] Codecov: Getting branch ref...
[2021-03-16T00:45:45.163Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-03-16T00:45:45.387Z] [INFO] Codecov: Sending data...
[2021-03-16T00:45:45.924Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2021-03-16T00:45:46.015Z] /tmp/.npm-global/bin/tap-xunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-03-16T00:45:46.015Z] /tmp/.npm-global/bin/txunit -> /tmp/.npm-global/lib/node_modules/tap-xunit/bin/tap-xunit
[2021-03-16T00:45:46.015Z] + tap-xunit@2.4.1
[2021-03-16T00:45:46.015Z] added 21 packages from 21 contributors in 2.559s
[2021-03-16T00:45:46.124Z] Recording test results
[2021-03-16T00:45:46.325Z] + bash codecov.sh
[2021-03-16T00:45:46.325Z] 
[2021-03-16T00:45:46.325Z]   _____          _
[2021-03-16T00:45:46.325Z]  / ____|        | |
[2021-03-16T00:45:46.325Z] | |     ___   __| | ___  ___ _____   __
[2021-03-16T00:45:46.325Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2021-03-16T00:45:46.325Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2021-03-16T00:45:46.325Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2021-03-16T00:45:46.325Z]                               Bash-20210309-2b87ace
[2021-03-16T00:45:46.325Z] 
[2021-03-16T00:45:46.325Z] 
[2021-03-16T00:45:46.325Z] ==> git version 2.17.1 found
[2021-03-16T00:45:46.325Z] ==> curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3
[2021-03-16T00:45:46.325Z] Release-Date: 2018-01-24
[2021-03-16T00:45:46.325Z] Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
[2021-03-16T00:45:46.325Z] Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
[2021-03-16T00:45:46.325Z] ==> Jenkins CI detected.
[2021-03-16T00:45:46.325Z]     project root: .
[2021-03-16T00:45:46.325Z] --> token set from env
[2021-03-16T00:45:46.325Z]     Yaml found at: .codecov.yml
[2021-03-16T00:45:46.325Z] ==> Running gcov in . (disable via -X gcov)
[2021-03-16T00:45:46.587Z] ==> Python coveragepy not found
[2021-03-16T00:45:46.587Z] ==> Searching for coverage reports in:
[2021-03-16T00:45:46.587Z]     + .
[2021-03-16T00:45:46.587Z] --> No coverage report found.
[2021-03-16T00:45:46.587Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2021-03-16T00:45:47.095Z] [Checks API] No suitable checks publisher found.
[2021-03-16T00:45:47.241Z] Running in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001/src/github.com/elastic/apm-agent-nodejs
[2021-03-16T00:45:47.347Z] Failed in branch Node.js-8-async-hooks-false
[2021-03-16T00:45:47.363Z] [INFO] Codecov: Getting branch ref...
[2021-03-16T00:45:47.457Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-03-16T00:45:47.684Z] [INFO] Codecov: Sending data...
[2021-03-16T00:45:48.233Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2021-03-16T00:45:48.640Z] + bash codecov.sh
[2021-03-16T00:45:48.640Z] 
[2021-03-16T00:45:48.640Z]   _____          _
[2021-03-16T00:45:48.640Z]  / ____|        | |
[2021-03-16T00:45:48.640Z] | |     ___   __| | ___  ___ _____   __
[2021-03-16T00:45:48.640Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2021-03-16T00:45:48.640Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2021-03-16T00:45:48.640Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2021-03-16T00:45:48.640Z]                               Bash-20210309-2b87ace
[2021-03-16T00:45:48.640Z] 
[2021-03-16T00:45:48.640Z] 
[2021-03-16T00:45:48.640Z] ==> git version 2.17.1 found
[2021-03-16T00:45:48.640Z] ==> curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3
[2021-03-16T00:45:48.640Z] Release-Date: 2018-01-24
[2021-03-16T00:45:48.640Z] Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
[2021-03-16T00:45:48.640Z] Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL 
[2021-03-16T00:45:48.640Z] ==> Jenkins CI detected.
[2021-03-16T00:45:48.640Z]     project root: .
[2021-03-16T00:45:48.640Z] --> token set from env
[2021-03-16T00:45:48.640Z]     Yaml found at: .codecov.yml
[2021-03-16T00:45:48.641Z] ==> Running gcov in . (disable via -X gcov)
[2021-03-16T00:45:48.903Z] ==> Python coveragepy not found
[2021-03-16T00:45:48.903Z] ==> Searching for coverage reports in:
[2021-03-16T00:45:48.903Z]     + .
[2021-03-16T00:45:48.903Z] --> No coverage report found.
[2021-03-16T00:45:48.903Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2021-03-16T00:45:49.317Z] Failed in branch Node.js-8
[2021-03-16T00:45:50.122Z] [INFO] getVaultSecret: Getting secrets
[2021-03-16T00:45:50.311Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-03-16T00:45:52.227Z] Stage "TAV Test" skipped due to earlier failure(s)
[2021-03-16T00:45:52.351Z] Stage "Edge Test" skipped due to earlier failure(s)
[2021-03-16T00:45:52.542Z] Stage "Nightly Test" skipped due to earlier failure(s)
[2021-03-16T00:45:52.569Z] Stage "Nightly Test - No async hooks" skipped due to earlier failure(s)
[2021-03-16T00:45:52.596Z] Stage "RC Test" skipped due to earlier failure(s)
[2021-03-16T00:45:52.622Z] Stage "RC Test - No async hooks" skipped due to earlier failure(s)
[2021-03-16T00:45:52.850Z] Failed in branch Nightly Test
[2021-03-16T00:45:52.880Z] Failed in branch Nightly Test - No async hooks
[2021-03-16T00:45:52.910Z] Failed in branch RC Test
[2021-03-16T00:45:52.939Z] Failed in branch RC Test - No async hooks
[2021-03-16T00:45:53.217Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-03-16T00:45:53.344Z] Stage "Release" skipped due to earlier failure(s)
[2021-03-16T00:45:53.394Z] Stage "Release" skipped due to earlier failure(s)
[2021-03-16T00:45:53.606Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-03-16T00:45:54.026Z] Running on Jenkins in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001
[2021-03-16T00:45:54.132Z] [INFO] getVaultSecret: Getting secrets
[2021-03-16T00:45:54.306Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-03-16T00:45:55.360Z] + chmod 755 generate-build-data.sh
[2021-03-16T00:45:55.360Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/1 FAILURE 606581
[2021-03-16T00:45:55.610Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/1/steps/?limit=10000 -o steps-info.json
[2021-03-16T00:45:56.161Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-03-16T00:45:56.411Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

trentm commented:
https://gist.github.com/trentm/4bafafb24aabc37f55b571858c972f27 demos the various logging configs.


# SCENARIO=0 logging capabilities before this change

```
% SCENARIO=0 node demo-apm-agent-pino-logging.js
could not extract route name from request { foo: 'bar', answer: 42 }
hi Bob
```

# SCENARIO=1 default ecs-logging format

```
% SCENARIO=1 node demo-apm-agent-pino-logging.js
{"log.level":"info","@timestamp":"2021-03-16T00:38:01.485Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"message":"could not extract route name from request {\"foo\":\"bar\",\"answer\":42}"}
{"log.level":"warn","@timestamp":"2021-03-16T00:38:01.486Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"message":"hi"}

{"log.level":"warn","@timestamp":"2021-03-16T00:38:01.487Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"error":{"type":"Error","message":"boom","stack_trace":"Error: boom\n    at Object.<anonymous> (/Users/trentm/tmp/p2/demo-apm-agent-pino-logging.js:106:20)\n    at Module._compile (internal/modules/cjs/loader.js:778:30)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)\n    at Module.load (internal/modules/cjs/loader.js:653:32)\n    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)\n    at Function.Module._load (internal/modules/cjs/loader.js:585:3)\n    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)\n    at startup (internal/bootstrap/node.js:283:19)\n    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)"},"message":"oops"}
{"log.level":"info","@timestamp":"2021-03-16T00:38:01.487Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"span":{"id":1},"a":1,"message":"started a span"}
{"log.level":"info","@timestamp":"2021-03-16T00:38:01.488Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"module":"this-is-the-client","foo":"bar","message":"hi from the http client"}
```

which can be pretty-printed (and filtered) with my fledgling [`ecslog` tool](https://github.com/trentm/go-ecslog):

![Screen Shot 2021-03-15 at 5 38 38 PM](https://user-images.githubusercontent.com/46866/111238945-6715f880-85b5-11eb-9304-b1c990e57d5e.png)


# SCENARIO=2 a custom pino logger

```
% SCENARIO=2 node demo-apm-agent-pino-logging.js
{"level":30,"time":1615855276596,"pid":22936,"hostname":"pink.local","name":"myLogger","msg":"could not extract route name from request {\"foo\":\"bar\",\"answer\":42}"}
{"level":40,"time":1615855276597,"pid":22936,"hostname":"pink.local","name":"myLogger","msg":"hi"}

{"level":40,"time":1615855276598,"pid":22936,"hostname":"pink.local","name":"myLogger","err":{"type":"Error","message":"boom","stack":"Error: boom\n    at Object.<anonymous> (/Users/trentm/tmp/p2/demo-apm-agent-pino-logging.js:106:20)\n    at Module._compile (internal/modules/cjs/loader.js:778:30)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)\n    at Module.load (internal/modules/cjs/loader.js:653:32)\n    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)\n    at Function.Module._load (internal/modules/cjs/loader.js:585:3)\n    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)\n    at startup (internal/bootstrap/node.js:283:19)\n    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)"},"msg":"oops"}
{"level":30,"time":1615855276598,"pid":22936,"hostname":"pink.local","name":"myLogger","span":{"id":1},"a":1,"msg":"started a span"}
{"level":30,"time":1615855276598,"pid":22936,"hostname":"pink.local","name":"myLogger","module":"this-is-the-client","foo":"bar","msg":"hi from the http client"}
```

This provides full fidelity. I.e. structured fields (like `a` above), serializers (like `err` and `span`), and child loggers are supported.

This can be pretty-printed with [pino-pretty](https://github.com/pinojs/pino-pretty)

# SCENARIO=3 a custom logger using console-log-level

A quick way to get the old logging format back.

```
% SCENARIO=3 node demo-apm-agent-pino-logging.js
could not extract route name from request {"foo":"bar","answer":42}
hi

oops
started a span
hi from the http client
```

Basically it works as before, but we can use the pino logger API in the agent code without breakage. The user of the custom logger should loses out on structured fields that we add.

# SCENARIO=4 using a bunyan logger

```
% SCENARIO=4 node demo-apm-agent-pino-logging.js
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":30,"msg":"could not extract route name from request {\"foo\":\"bar\",\"answer\":42}","time":"2021-03-16T00:44:29.624Z","v":0}
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":40,"msg":"hi","time":"2021-03-16T00:44:29.625Z","v":0}

{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":40,"msg":"oops","time":"2021-03-16T00:44:29.626Z","v":0}
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":30,"msg":"started a span","time":"2021-03-16T00:44:29.626Z","v":0}
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":30,"msg":"hi from the http client","time":"2021-03-16T00:44:29.626Z","v":0}
```

I only really included this because our current docs mention using a Bunyan custom logger. It works, but loses fidelity for new structured logging. (The bunyan API is similar enough to pino -- pino was based on it -- that we *could* keep fidelity with a custom Bunyan logger. However I propose that we do *not* sign up for that long term compat game.)

trentm commented:
# format change impact on user

The default log level is "info", so with the change to pino format logging,
all info, warn, and error log statements will will be different for the default
case.

For example:

```
# Before
APM Server transport error (ECONNREFUSED): connect ECONNREFUSED 127.0.0.1:8200

# After (newline added to attempt simulate typical wrapping in the console)
{"log.level":"error","@timestamp":"2021-03-19T00:21:17.571Z","log":{"logger":"elastic-apm-node"},
"ecs":{"version":"1.6.0"},"message":"APM Server transport error (ECONNREFUSED): connect ECONNREFUSED 127.0.0.1:8200"}
```

Or, more nicely if/when `ecslog` gets used:

![Screen Shot 2021-03-18 at 5 24 26 PM](https://user-images.githubusercontent.com/46866/111714477-f538eb80-880e-11eb-9421-df69d73e26f0.png)



Here are all the current `logger.info` cases:

```
% rg logger.info
lib/agent.js
418:      agent.logger.info('Sending error to Elastic APM: %o', { id })

lib/config.js
320:                agent.logger.info(`Remote config success. Updated logger with new logLevel: ${value}`)
322:              agent.logger.info(`Remote config success. Updated ${key}: ${value}`)
```

and `logger.warn` cases:

```
% rg logger.warn
lib/agent.js
    -   const err = new Error('cannot flush agent before it is started')
465:    this.logger.warn({ err }, err.message)

lib/instrumentation/http-shared.js
198:        agent.logger.warn('unable to identify http.ClientRequest url %o', { id: id })

lib/instrumentation/generic-span.js
101:    this._agent.logger.warn('Illegal characters used in tag key: %s', key)

lib/instrumentation/modules/fastify.js
74:    agent.logger.warn('Elastic APM cannot automaticaly capture errors on this verison of Fastify. Upgrade to version 2.0.0 or later.')

lib/config.js
309:            agent.logger.warn(`Remote config failure. Unsupported config names: ${unknown.join(', ')}`)
427:      logger.warn('Invalid "transactionSampleRate" config value %s, falling back to default %s',
451:      logger.warn('Invalid "cloudProvider" config value %s, falling back to default %s',
614:      logger.warn('unrecognized boolean value "%s" for "%s"', value, key)

lib/logging.js
159:    logger.warn('unknown log levelName "%s": cannot setLogLevel', levelName)
```

and `logger.error` cases:

```
% rg logger.error
lib/agent.js
164:    this.logger.error('Elastic APM isn\'t correctly configured: Missing serviceName')
168:    this.logger.error('Elastic APM isn\'t correctly configured: serviceName "%s" contains invalid characters! (allowed: a-z, A-Z, 0-9, _, -, <space>)', this._conf.serviceName)
172:    this.logger.error('Elastic APM isn\'t correctly configured: serverUrl "%s" contains an invalid port! (allowed: 1-65535)', this._conf.serverUrl)
257:    this.logger.error('Can\'t add filter of type %s', typeof fn)
266:    this.logger.error('Can\'t add filter of type %s', typeof fn)
275:    this.logger.error('Can\'t add filter of type %s', typeof fn)
483:    this.logger.error('Can\'t add callback of type %s', typeof callback)

lib/lambda.js
51:        if (err) agent.logger.error('Flush error:', err.message)

lib/instrumentation/transaction.js
169:    this._agent.logger.error('cannot encode un-ended transaction: %o', { trans: this.id, parent: this.parentId, trace: this.traceId })

lib/config.js
328:          agent.logger.error('APM Server transport error: %s', err.stack)
360:          agent.logger.error(msg)
650:      opts.logger.error('Elastic APM initialization error: Can\'t read server CA cert file %s (%s)', opts.serverCaCertFile, err.message)

lib/instrumentation/http-shared.js
228:            agent.logger.error('Could not set destination context:', e.message)

lib/instrumentation/index.js
101:      this._agent.logger.error('Invalid patch handler type:', type)
199:        this._agent.logger.error('Invalid patch handler type "%s" for module "%s"', type, name)
229:        agent.logger.error('error encoding span %o', { span: span.id, parent: span.parentId, trace: span.traceId, name: span.name, type: span.type, error: err.message })

lib/cloud-metadata/callback-coordination.js
37:          this.logger.error('metadata requests did not finish, possible deadlock')
77:      this.logger.error('Can not schedule callback, already started')

lib/instrumentation/modules/http2.js
192:          agent.logger.error('Could not set destination context:', e.message)

lib/instrumentation/modules/cassandra-driver.js
61:          agent.logger.error('unable to identify span exit point for cassandra-driver')
117:          agent.logger.error('unable to identify span exit point for cassandra-driver')
156:          agent.logger.error('unable to identify span exit point for cassandra-driver')
190:          agent.logger.error('unable to identify span exit point for cassandra-driver')
```

trentm commented:
An open question is whether this logging format change is a negative impact on the user that we need/want/care to mitigate.  True the log *message* is by default (i.e. without `ecslog` usage) less visible, but it is still there. The ECS log output includes a "log.level" which can give the user some idea of severity that wasn't communicated before. I think that'll be helpful, for example, to change the "APM Server transport error" to be a *warning* in a subsequent change. The "log.logger=elastic-apm-node" also can help disambiguate in a noisy log that the message is from the APM agent, vs. from some part of a large app.

My inclination is this format change is fine, i.e. not something we need to mitigate for users looking at the log output on the console. However, I'm biased. @astorm do you have an opinion on this?

astorm commented:
> My inclination is this format change is fine, i.e. not something we need to mitigate for users looking at the log output on the console.

That seems reasonable.  While the format has changed/become more complex, the `message` property still contains the same text that a user might be visually-scanning or `grep`-ing for.  No need for mitigation other than mentioning in the CHANGELOG that the log format has changed. 

@trentm

astorm commented on lib/agent.js:
```diff
@@ -459,7 +459,9 @@ Agent.prototype.flush = function (cb) {
-    this.logger.warn(new Error('cannot flush agent before it is started'))
+    // Log an *err* to provide a stack for the user.
+    const err = new Error('cannot flush agent before it is started')
+    this.logger.warn({ err }, err.message)
```
This is interesting to me -- I didn't know about pino's optional `mergingObject` parameter/argument until [researching this syntax](https://getpino.io/#/docs/api?id=trace). (nothing to see/do here, carry on)

trentm replied:
It originates from https://github.com/trentm/node-bunyan#log-method-api but there are some minor differences.
- pino dropped the `log.info()`-with-no-args call signature for the more traditional log4j `log.isLevelEnabled(level)` (https://getpino.io/#/docs/api?id=islevelenabled) which was a good call.
- Bunyan treats `log.info(err)` and `log.info({err: err})` the same. Pino does not. The form that logs the error with the `err` field name is used by `@elastic/ecs-pino-format` module to format the error fields into ECS format, so we should bias to using that form.
- Bunyan will use `err.message` for the "message" field with `log.info({err: err})`. Pino will leave the message empty. Arguably pino's is cleaner, but for the (rare) case of an APM Agent user specifying a custom `logger`, they will only get the "message" field, so it is nice to have *some* message there.

astorm reviewed (dismissed):
Looks like a welcome addition to the agent.  We get ourselves out of the `console-log-level`  business, get some nice overlap with our `ecs-format-pino`, and our users get all the advantages of using one of Node's mature defacto logging libraries.  

Approving, pending CI passing after the additional work on the "placeholder %s", fixes

trentm commented on lib/logging.js:
```diff
@@ -29,68 +28,142 @@
-  logger[LOGGER_IS_OURS_SYM] = true // used for isLoggerCustom()
-  logger.level = logApiLevel // used for testing
+  if (customLogger) {
+    // Is this a pino logger? If so, it supports to API the agent requires and
```
Yes. Typo. Thanks. I'll fix that.

trentm replied:
commit 174f977

bmorelli25 commented on docs/configuration.asciidoc:
```diff
@@ -431,34 +431,66 @@ require('elastic-apm-node').start({
+* *Env:* `ELASTIC_APM_LOGGER=false` to _ignore_ a custom logger
+
+By default, the APM agent logs to stdout in
+https://www.elastic.co/guide/en/ecs-logging/overview/current/intro.html[ecs-logging]
```
```suggestion
{ecs-logging-ref}/intro.html[ecs-logging]
```

bmorelli25 commented on CHANGELOG.asciidoc:
```diff
@@ -36,6 +36,32 @@ Notes:
 
+* The APM agent's own internal logging now uses structured JSON logging using
+  the https://getpino.io/#/docs/api?id=logger[pino API], and formatted in
+  https://www.elastic.co/guide/en/ecs-logging/overview/current/intro.html[ecs-logging]
```
```suggestion
  {ecs-logging-ref}/intro.html[ecs-logging]
```

bmorelli25 commented on docs/troubleshooting.asciidoc:
```diff
@@ -114,9 +114,28 @@ Please create a new topic in the https://discuss.elastic.co/c/apm[Elastic APM di
+----
+
+If you are capturing debugging output for Elastic support, for help on the
+Elastic forums, or for a GitHub issue, please *please upload the complete
```
Not sure if the double please was intended here

```suggestion
Elastic forums, or for a GitHub issue, *please upload the complete
```

trentm replied:
Ha ha, no it wasn't.

bmorelli25 reviewed (commented):
This looks really good. Just a few suggestions.

trentm commented:
I'll assume rubberstamp re-approve by Alan and Brandon.

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/16//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/16//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/16//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/16//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #2001 updated`


* Reason: `Aborted from #17`


* Start Time: 2021-03-23T16:26:39.162+0000


* Duration: 183 min 33 sec


* Commit: 5dbb65f655dfd5681e711115e723eb3f32edb4b3



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 4  |
| Passed       | 1846  |
| Skipped      | 0 |
| Total        | 1850   |


#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2001/buildTimeGraph/png)


![Image of Tests](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2001/16/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![4](https://img.shields.io/badge/4%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/16//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Test / Node.js-14-async-hooks-false / [empty] – *-output-junit.xml`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Test report file /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001/src/github.com/elastic/apm-agent-nodejs/*-output-junit.xml was length 0 
 ```
</p></details>
</ul>
    
##### `Test / Node.js-14 / [empty] – *-output-junit.xml`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Test report file /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001/src/github.com/elastic/apm-agent-nodejs/*-output-junit.xml was length 0 
 ```
</p></details>
</ul>
    
##### `Test / Node.js-14.0-async-hooks-false / [empty] – *-output-junit.xml`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Test report file /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001/src/github.com/elastic/apm-agent-nodejs/*-output-junit.xml was length 0 
 ```
</p></details>
</ul>
    
##### `Test / Node.js-14.0 / [empty] – *-output-junit.xml`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Test report file /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001/src/github.com/elastic/apm-agent-nodejs/*-output-junit.xml was length 0 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-2001/detail/PR-2001/16//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Shell Script`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/16/steps/615/log/?start=0">here</a></li>
<li>Description: <code>      docker run --rm         -v $(pwd):/usr/src/app         -w /usr/src/app         -u $(id -u):$(id -g)         node:12-alpine         sh -c 'export HOME=/tmp ; mkdir ~/.npm-global; npm config set prefix ~/.npm-global ; npm install tap-xunit -g ; for i in "*-output.tap" ; do cat ${i} | /tmp/.npm-global/bin/tap-xunit --package="Node.js" > ${i%.*}-junit.xml ; done'    </code></l1>
</ul>
  
  
  
  
##### `Shell Script`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/16/steps/621/log/?start=0">here</a></li>
<li>Description: <code>      docker run --rm         -v $(pwd):/usr/src/app         -w /usr/src/app         -u $(id -u):$(id -g)         node:12-alpine         sh -c 'export HOME=/tmp ; mkdir ~/.npm-global; npm config set prefix ~/.npm-global ; npm install tap-xunit -g ; for i in "*-output.tap" ; do cat ${i} | /tmp/.npm-global/bin/tap-xunit --package="Node.js" > ${i%.*}-junit.xml ; done'    </code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2001/runs/16/steps/337/log/?start=0">here</a></li>
<li>Description: <code>org.jenkinsci.plugins.workflow.steps.FlowInterruptedException</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-03-23T16:41:50.651Z] # ok
[2021-03-23T16:41:50.651Z] 
[2021-03-23T16:41:50.651Z] running (cwd: ./): node --unhandled-rejections=strict test\uncaught-exceptions\log-uncaught-exceptions-on.js
[2021-03-23T16:41:51.226Z] TAP version 13
[2021-03-23T16:41:51.226Z] # should capture and log uncaught exceptions by default
[2021-03-23T16:41:51.226Z] ok 1 should log the error to STDERR
[2021-03-23T16:41:51.226Z] Error: foo
[2021-03-23T16:41:51.226Z]     at Test.<anonymous> (C:\Users\jenkins\workspace\ejs_apm-agent-nodejs-mbp_PR-2001\src\github.com\elastic\apm-agent-nodejs\test\uncaught-exceptions\log-uncaught-exceptions-on.js:43:23)
[2021-03-23T16:41:51.226Z]     at Test.bound [as _cb] (C:\Users\jenkins\workspace\ejs_apm-agent-nodejs-mbp_PR-2001\src\github.com\elastic\apm-agent-nodejs\node_modules\tape\lib\test.js:91:32)
[2021-03-23T16:41:51.226Z]     at Test.run (C:\Users\jenkins\workspace\ejs_apm-agent-nodejs-mbp_PR-2001\src\github.com\elastic\apm-agent-nodejs\node_modules\tape\lib\test.js:108:31)
[2021-03-23T16:41:51.227Z]     at Test.bound [as run] (C:\Users\jenkins\workspace\ejs_apm-agent-nodejs-mbp_PR-2001\src\github.com\elastic\apm-agent-nodejs\node_modules\tape\lib\test.js:91:32)
[2021-03-23T16:41:51.227Z]     at Immediate.next (C:\Users\jenkins\workspace\ejs_apm-agent-nodejs-mbp_PR-2001\src\github.com\elastic\apm-agent-nodejs\node_modules\tape\lib\results.js:89:19)
[2021-03-23T16:41:51.227Z]     at Immediate.elasticAPMCallbackWrapper (C:\Users\jenkins\workspace\ejs_apm-agent-nodejs-mbp_PR-2001\src\github.com\elastic\apm-agent-nodejs\lib\instrumentation\index.js:328:27)
[2021-03-23T16:41:51.227Z]     at processImmediate (internal/timers.js:461:21)
[2021-03-23T16:41:51.227Z] ok 2 should capture the error
[2021-03-23T16:41:51.487Z] {"log.level":"info","@timestamp":"2021-03-23T16:41:50.910Z","log":{"logger":"elastic-apm-node"},"ecs":{"version":"1.6.0"},"message":"Sending error to Elastic APM: {\"id\":\"f3fcce17bc6a60187cd7db7917954178\"}"}
[2021-03-23T16:41:51.487Z] ok 3 should be strictly equal
[2021-03-23T16:41:51.487Z] ok 4 should be strictly equal
[2021-03-23T16:41:51.487Z] ok 5 should be strictly equal
[2021-03-23T16:41:51.487Z] ok 6 should have a stack trace
[2021-03-23T16:41:51.487Z] ok 7 stack trace should contain frames
[2021-03-23T16:41:51.487Z] ok 8 top frame should be this file
[2021-03-23T16:41:51.487Z] 
[2021-03-23T16:41:51.487Z] 1..8
[2021-03-23T16:41:51.487Z] # tests 8
[2021-03-23T16:41:51.487Z] # pass  8
[2021-03-23T16:41:51.487Z] 
[2021-03-23T16:41:51.487Z] # ok
[2021-03-23T16:41:51.487Z] 
[2021-03-23T16:41:51.515Z] JUnit archiving no yet in place
[2021-03-23T16:41:51.876Z] + .ci/scripts/windows/stop-test.sh 12
[2021-03-23T16:41:51.876Z] + NODE_VERSION=12
[2021-03-23T16:41:51.876Z] + NODE_VERSION=12
[2021-03-23T16:41:51.876Z] + docker-compose --no-ansi -f .ci/docker/docker-compose-all.yml logs --timestamps
[2021-03-23T16:41:55.225Z] + NODE_VERSION=12
[2021-03-23T16:41:55.225Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml down -v
[2021-03-23T16:41:55.798Z] Stopping docker_redis_1         ... 
[2021-03-23T16:41:55.798Z] Stopping docker_cassandra_1     ... 
[2021-03-23T16:41:55.798Z] Stopping docker_mssql_1         ... 
[2021-03-23T16:41:55.798Z] Stopping docker_memcached_1     ... 
[2021-03-23T16:41:55.798Z] Stopping docker_elasticsearch_1 ... 
[2021-03-23T16:41:55.798Z] Stopping docker_postgres_1      ... 
[2021-03-23T16:41:55.798Z] Stopping docker_mysql_1         ... 
[2021-03-23T16:41:55.798Z] Stopping docker_mongodb_1       ... 
[2021-03-23T16:41:56.376Z] Stopping docker_redis_1         ... done
[2021-03-23T16:41:56.376Z] Stopping docker_mongodb_1       ... done
[2021-03-23T16:41:56.636Z] Stopping docker_postgres_1      ... done
[2021-03-23T16:41:56.636Z] Stopping docker_elasticsearch_1 ... done
[2021-03-23T16:41:56.897Z] Stopping docker_mssql_1         ... done
[2021-03-23T16:41:57.842Z] Stopping docker_mysql_1         ... done
[2021-03-23T16:41:58.106Z] Stopping docker_memcached_1     ... done
[2021-03-23T16:41:58.689Z] Stopping docker_cassandra_1     ... done
[2021-03-23T16:41:58.689Z] Removing docker_node_tests_1    ... 
[2021-03-23T16:41:58.689Z] Removing docker_redis_1         ... 
[2021-03-23T16:41:58.689Z] Removing docker_cassandra_1     ... 
[2021-03-23T16:41:58.689Z] Removing docker_mssql_1         ... 
[2021-03-23T16:41:58.689Z] Removing docker_memcached_1     ... 
[2021-03-23T16:41:58.689Z] Removing docker_elasticsearch_1 ... 
[2021-03-23T16:41:58.689Z] Removing docker_postgres_1      ... 
[2021-03-23T16:41:58.689Z] Removing docker_mysql_1         ... 
[2021-03-23T16:41:58.689Z] Removing docker_mongodb_1       ... 
[2021-03-23T16:41:58.689Z] Removing docker_cassandra_1     ... done
[2021-03-23T16:41:58.689Z] Removing docker_mssql_1         ... done
[2021-03-23T16:41:58.689Z] Removing docker_mysql_1         ... done
[2021-03-23T16:41:58.689Z] Removing docker_mongodb_1       ... done
[2021-03-23T16:41:58.948Z] Removing docker_memcached_1     ... done
[2021-03-23T16:41:58.948Z] Removing docker_redis_1         ... done
[2021-03-23T16:41:58.949Z] Removing docker_postgres_1      ... done
[2021-03-23T16:41:58.949Z] Removing docker_elasticsearch_1 ... done
[2021-03-23T16:41:59.519Z] Removing docker_node_tests_1    ... done
[2021-03-23T19:29:03.825Z] Cancelling nested steps due to timeout
[2021-03-23T19:29:03.828Z] Could not connect to apm-ci-immutable-ubuntu-1804-1616516923062926497 to send interrupt signal to process
[2021-03-23T19:29:03.830Z] Could not connect to apm-ci-immutable-ubuntu-1804-1616516923062606428 to send interrupt signal to process
[2021-03-23T19:29:03.841Z] Sleeping for 5 sec
[2021-03-23T19:29:03.846Z] Sleeping for 5 sec
[2021-03-23T19:29:08.985Z] Failed in branch Node.js-13
[2021-03-23T19:29:09.008Z] Failed in branch Node.js-12
[2021-03-23T19:29:09.101Z] [INFO] getVaultSecret: Getting secrets
[2021-03-23T19:29:09.308Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-03-23T19:29:10.631Z] Stage "TAV Test" skipped due to earlier failure(s)
[2021-03-23T19:29:10.675Z] Stage "Edge Test" skipped due to earlier failure(s)
[2021-03-23T19:29:10.735Z] Stage "Nightly Test" skipped due to earlier failure(s)
[2021-03-23T19:29:10.737Z] Stage "Nightly Test - No async hooks" skipped due to earlier failure(s)
[2021-03-23T19:29:10.739Z] Stage "RC Test" skipped due to earlier failure(s)
[2021-03-23T19:29:10.741Z] Stage "RC Test - No async hooks" skipped due to earlier failure(s)
[2021-03-23T19:29:10.793Z] Failed in branch Nightly Test
[2021-03-23T19:29:10.795Z] Failed in branch Nightly Test - No async hooks
[2021-03-23T19:29:10.797Z] Failed in branch RC Test
[2021-03-23T19:29:10.799Z] Failed in branch RC Test - No async hooks
[2021-03-23T19:29:10.868Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-03-23T19:29:10.918Z] Stage "Release" skipped due to earlier failure(s)
[2021-03-23T19:29:10.938Z] Stage "Release" skipped due to earlier failure(s)
[2021-03-23T19:29:11.004Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-03-23T19:29:11.334Z] Running on Jenkins in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-2001
[2021-03-23T19:29:11.380Z] [INFO] getVaultSecret: Getting secrets
[2021-03-23T19:29:11.573Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-03-23T19:29:12.276Z] + chmod 755 generate-build-data.sh
[2021-03-23T19:29:12.276Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/16 ABORTED 10952847
[2021-03-23T19:29:12.827Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/16/steps/?limit=10000 -o steps-info.json
[2021-03-23T19:29:13.378Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2001/runs/16/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->
